File: sdks/node/src/api/core/core_mapper.ts
import type { CyanGlobReq, CyanPluginReq, CyanProcessorReq, CyanReq } from './cyan_req.js';
import type { Cyan, CyanGlob, CyanPlugin, CyanProcessor } from '../../domain/core/cyan.js';
import { GlobType } from '../../domain/core/cyan.js';
import type { CyanGlobRes, CyanPluginRes, CyanProcessorRes, CyanRes } from './cyan_res.js';
import type { CheckboxQ, ConfirmQ, DateQ, PasswordQ, Question, SelectQ, TextQ } from '../../domain/core/question.js';

import type {
  CheckboxQuestionRes,
  ConfirmQuestionRes,
  DateQuestionRes,
  PasswordQuestionRes,
  QuestionRes,
  SelectQuestionRes,
  TextQuestionRes,
} from './question_res.js';
import { type Answer, isBoolAnswer, isStringAnswer, isStringArrayAnswer } from '../../domain/core/answer.js';
import type { AnswerRes } from './answer_res.js';
import { type AnswerReq, isBoolAnswerReq, isStringAnswerReq } from './answer_req.js';
import { QuestionType } from '../../domain/core/question.js';

export class CyanMapper {
  static pluginReqToDomain(req: CyanPluginReq): CyanPlugin {
    return {
      name: req.name,
      config: req.config,
    };
  }

  static globReqToDomain(req: CyanGlobReq): CyanGlob {
    const type = (() => {
      switch (req.type) {
        case 'template':
          return GlobType.Template;
        case 'copy':
          return GlobType.Copy;
        default:
          throw new Error(`Invalid req.type: ${req.type}`);
      }
    })();
    return {
      root: req.root,
      glob: req.glob,
      exclude: req.exclude,
      type,
    };
  }

  static processorReqToDomain(req: CyanProcessorReq): CyanProcessor {
    return {
      name: req.name,
      config: req.config,
      files: req.files.map(x => CyanMapper.globReqToDomain(x)),
    };
  }

  static cyanReqToDomain(req: CyanReq): Cyan {
    return {
      processors: req.processors.map(x => CyanMapper.processorReqToDomain(x)),
      plugins: req.plugins.map(x => CyanMapper.pluginReqToDomain(x)),
    };
  }

  static cyanToResp(data: Cyan): CyanRes {
    return {
      processors: data.processors.map(x => CyanMapper.processorToResp(x)),
      plugins: data.plugins.map(x => CyanMapper.pluginToResp(x)),
    };
  }

  static processorToResp(data: CyanProcessor): CyanProcessorRes {
    return {
      name: data.name,
      config: data.config,
      files: data.files.map(x => CyanMapper.globToResp(x)),
    };
  }

  static pluginToResp(data: CyanPlugin): CyanPluginRes {
    return {
      name: data.name,
      config: data.config,
    };
  }

  static globTypeToResp(type: GlobType): string {
    switch (type) {
      case GlobType.Template:
        return 'template';
      case GlobType.Copy:
        return 'copy';
      default:
        throw new Error(`Invalid GlobType: ${type}`);
    }
  }

  static globToResp(data: CyanGlob): CyanGlobRes {
    return {
      root: data.root,
      glob: data.glob,
      exclude: data.exclude,
      type: CyanMapper.globTypeToResp(data.type),
    };
  }
}

export class QuestionMapper {
  static questionToResp(q: Question): QuestionRes {
    switch (q.type) {
      case QuestionType.Select:
        return QuestionMapper.selectToResp(q);
      case QuestionType.Text:
        return QuestionMapper.textToResp(q);
      case QuestionType.Password:
        return QuestionMapper.passwordToResp(q);
      case QuestionType.DateSelect:
        return QuestionMapper.dateToResp(q);
      case QuestionType.Confirm:
        return QuestionMapper.confirmToResp(q);
      case QuestionType.Checkbox:
        return QuestionMapper.checkboxToResp(q);
      default:
        throw new Error(`Invalid question type: ${q}`);
    }
  }

  static confirmToResp(q: ConfirmQ): ConfirmQuestionRes {
    return {
      default: q.default ?? null,
      message: q.message,
      id: q.id,
      errorMessage: q.errorMessage ?? null,
      desc: q.desc ?? null,
      type: 'confirm',
    };
  }

  static checkboxToResp(q: CheckboxQ): CheckboxQuestionRes {
    return {
      message: q.message,
      id: q.id,
      desc: q.desc ?? null,
      options: q.options,
      type: 'checkbox',
    };
  }

  static selectToResp(q: SelectQ): SelectQuestionRes {
    return {
      message: q.message,
      id: q.id,
      desc: q.desc ?? null,
      options: q.options,
      type: 'select',
    };
  }

  static textToResp(q: TextQ): TextQuestionRes {
    return {
      message: q.message,
      id: q.id,
      desc: q.desc ?? null,
      default: q.default ?? null,
      initial: q.initial ?? null,
      type: 'text',
    };
  }

  static passwordToResp(q: PasswordQ): PasswordQuestionRes {
    return {
      message: q.message,
      id: q.id,
      desc: q.desc ?? null,
      confirmation: q.confirmation ?? null,
      type: 'password',
    };
  }

  static dateToResp(q: DateQ): DateQuestionRes {
    return {
      message: q.message,
      id: q.id,
      desc: q.desc ?? null,
      default: q.default ? q.default.toISOString().split('T')[0] : null,
      maxDate: q.maxDate ? q.maxDate.toISOString().split('T')[0] : null,
      minDate: q.minDate ? q.minDate.toISOString().split('T')[0] : null,
      type: 'date',
    };
  }
}

export class AnswerMapper {
  static toDomain(req: AnswerReq): Answer {
    if (isBoolAnswerReq(req)) {
      return { answer: req.answer };
    }
    if (isStringAnswerReq(req)) {
      return { answer: req.answer };
    }
    if (isStringArrayAnswer(req)) {
      return { answer: req.answer };
    }
    throw new Error(`Invalid answer request type: ${req}`);
  }

  static toResp(answer: Answer): AnswerRes {
    if (isBoolAnswer(answer)) {
      return { answer: answer.answer };
    }
    if (isStringAnswer(answer)) {
      return { answer: answer.answer };
    }
    if (isStringArrayAnswer(answer)) {
      return { answer: answer.answer };
    }
    throw new Error(`Invalid answer type: ${answer}`);
  }
}

File: sdks/dotnet/sulfone-helium/Api/Core/CoreMapper.cs
using System.Globalization;
using sulfone_helium.Domain.Core;
using sulfone_helium.Domain.Core.Questions;

namespace sulfone_helium.Api.Core;

public static class CyanMapper
{
    public static CyanPlugin ToDomain(this CyanPluginReq req)
    {
        return new CyanPlugin { Name = req.Name, Config = req.Config.ToDynamic() };
    }

    public static CyanGlob ToDomain(this CyanGlobReq req)
    {
        return new CyanGlob
        {
            Root = req.Root,
            Glob = req.Glob,
            Exclude = req.Exclude,
            Type = req.Type switch
            {
                "template" => GlobType.Template,
                "copy" => GlobType.Copy,
                _ => throw new ArgumentOutOfRangeException(nameof(req.Type), req.Type, null),
            },
        };
    }

    public static CyanProcessor ToDomain(this CyanProcessorReq req)
    {
        return new CyanProcessor
        {
            Name = req.Name,
            Config = req.Config.ToDynamic(),
            Files = req.Files.Select(x => x.ToDomain()).ToArray(),
        };
    }

    public static Cyan ToDomain(this CyanReq req)
    {
        return new Cyan
        {
            Processors = req.Processors.Select(x => x.ToDomain()).ToArray(),
            Plugins = req.Plugins.Select(x => x.ToDomain()).ToArray(),
        };
    }

    public static CyanRes ToResp(this Cyan data)
    {
        return new CyanRes
        {
            Processors = data.Processors.Select(x => x.ToResp()).ToArray(),
            Plugins = data.Plugins.Select(x => x.ToResp()).ToArray(),
        };
    }

    public static CyanProcessorRes ToResp(this CyanProcessor data)
    {
        return new CyanProcessorRes()
        {
            Name = data.Name,
            Config = Utility.ToDynamic(data.Config),
            Files = data.Files.Select(x => x.ToResp()).ToArray(),
        };
    }

    public static CyanPluginRes ToResp(this CyanPlugin data)
    {
        return new CyanPluginRes() { Name = data.Name, Config = Utility.ToDynamic(data.Config) };
    }

    public static string ToResp(this GlobType type)
    {
        return type switch
        {
            GlobType.Template => "template",
            GlobType.Copy => "copy",
            _ => throw new ArgumentOutOfRangeException(nameof(type), type, null),
        };
    }

    public static CyanGlobRes ToResp(this CyanGlob data)
    {
        return new CyanGlobRes()
        {
            Root = data.Root,
            Glob = data.Glob,
            Exclude = data.Exclude,
            Type = data.Type.ToResp(),
        };
    }
}

public static class QuestionMapper
{
    public static QuestionRes ToResp(this IQuestion q)
    {
        return q switch
        {
            CheckboxQ checkbox => checkbox.ToResp(),
            ConfirmQ confirm => confirm.ToResp(),
            DateQ date => date.ToResp(),
            PasswordQ password => password.ToResp(),
            SelectQ select => select.ToResp(),
            TextQ text => text.ToResp(),
            _ => throw new ArgumentOutOfRangeException(nameof(q)),
        };
    }

    public static ConfirmQuestionRes ToResp(this ConfirmQ q)
    {
        return new ConfirmQuestionRes()
        {
            Default = q.Default,
            Message = q.Message,
            Id = q.Id,
            ErrorMessage = q.ErrorMessage,
            Desc = q.Desc,
        };
    }

    public static CheckboxQuestionRes ToResp(this CheckboxQ q)
    {
        return new CheckboxQuestionRes()
        {
            Message = q.Message,
            Id = q.Id,
            Desc = q.Desc,
            Options = q.Options,
        };
    }

    public static SelectQuestionRes ToResp(this SelectQ q)
    {
        return new SelectQuestionRes()
        {
            Message = q.Message,
            Id = q.Id,
            Desc = q.Desc,
            Options = q.Options,
        };
    }

    public static TextQuestionRes ToResp(this TextQ q)
    {
        return new TextQuestionRes()
        {
            Message = q.Message,
            Id = q.Id,
            Desc = q.Desc,
            Default = q.Default,
            Initial = q.Initial,
        };
    }

    public static PasswordQuestionRes ToResp(this PasswordQ q)
    {
        return new PasswordQuestionRes()
        {
            Message = q.Message,
            Id = q.Id,
            Desc = q.Desc,
            Confirmation = q.Confirmation,
        };
    }

    public static DateQuestionRes ToResp(this DateQ q)
    {
        return new DateQuestionRes()
        {
            Message = q.Message,
            Id = q.Id,
            Desc = q.Desc,
            Default = q.Default?.ToString("o", CultureInfo.InvariantCulture),
            MaxDate = q.MaxDate?.ToString("o", CultureInfo.InvariantCulture),
            MinDate = q.MinDate?.ToString("o", CultureInfo.InvariantCulture),
        };
    }
}

public static class AnswerMapper
{
    public static IAnswer ToDomain(this AnswerReq req)
    {
        return req switch
        {
            BoolAnswerReq boolAnswer => new BoolAnswer(boolAnswer.Answer),
            StringAnswerReq stringAnswer => new StringAnswer(stringAnswer.Answer),
            StringArrayAnswerReq stringArrayAnswer => new StringArrayAnswer(
                stringArrayAnswer.Answer
            ),
            _ => throw new ArgumentOutOfRangeException(nameof(req)),
        };
    }

    public static AnswerRes ToResp(this IAnswer answer)
    {
        return answer switch
        {
            BoolAnswer boolAnswer => new BoolAnswerRes() { Answer = boolAnswer.Answer },
            StringAnswer stringAnswer => new StringAnswerRes() { Answer = stringAnswer.Answer },
            StringArrayAnswer stringArrayAnswer => new StringArrayAnswerRes()
            {
                Answer = stringArrayAnswer.Answer,
            },
            _ => throw new ArgumentOutOfRangeException(nameof(answer)),
        };
    }
}
